src/main/java/org/cryptomator/cloudaccess/MetadataCachingProviderDecorator.java

```diff
@@ -70,94 +70,98 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
-						cloudItemList.getItems().forEach(metadata -> itemMetadataCache.put(metadata.getPath(), CompletableFuture.completedFuture(metadata)));
-					}
-				});
+		try {
+			var folderKey = Map.entry(folder, pageToken);
+			return cachedItemListRequests.get(folderKey, () -> delegate.list(folder, pageToken).whenComplete((metadata, throwable) -> evict(folderKey)));
+		} catch (ExecutionException e) {
+			return CompletableFuture.failedFuture(e);
+		}
 	}
 
 	@Override
 	public CompletionStage<InputStream> read(CloudPath file, ProgressListener progressListener) {
-		return delegate.read(file, progressListener) //
-				.whenComplete((metadata, exception) -> {
-					if (exception != null) {
-						itemMetadataCache.invalidate(file);
-					}
-				});
+		return delegate.read(file, progressListener).whenComplete((metadata, exception) -> {
+			if (exception != null) {
+				evict(file);
+			}
+		});
 	}
 
 	@Override
 	public CompletionStage<InputStream> read(CloudPath file, long offset, long count, ProgressListener progressListener) {
-		return delegate.read(file, offset, count, progressListener) //
-				.whenComplete((inputStream, exception) -> {
-					if (exception != null) {
-						itemMetadataCache.invalidate(file);
-					}
-				});
+		return delegate.read(file, offset, count, progressListener).whenComplete((inputStream, exception) -> {
+			if (exception != null) {
+				evict(file);
+			}
+		});
 	}
 
 	@Override
 	public CompletionStage<Void> write(CloudPath file, boolean replace, InputStream data, long size, Optional<Instant> lastModified, ProgressListener progressListener) {
-		return delegate.write(file, replace, data, size, lastModified, progressListener) //
-				.whenComplete((nullReturn, exception) -> {
-					if (exception == null) {
-						itemMetadataCache.put(file, CompletableFuture.completedFuture(new CloudItemMetadata(file.getFileName().toString(), file, CloudItemType.FILE, lastModified, Optional.of(size))));
-						quotaCache.invalidateAll();
-					} else {
-						itemMetadataCache.invalidate(file);
-					}
-				});
+		return delegate.write(file, replace, data, size, lastModified, progressListener).whenComplete((nullReturn, exception) -> {
+			if (exception != null) {
+				evict(file);
+			}
+		});
 	}
 
 	@Override
 	public CompletionStage<CloudPath> createFolder(CloudPath folder) {
-		return delegate.createFolder(folder) //
-				.whenComplete((metadata, exception) -> {
-					itemMetadataCache.invalidate(folder);
-					quotaCache.invalidateAll();
-				});
+		return delegate.createFolder(folder).whenComplete((metadata, exception) -> evict(folder));
 	}
 
 	@Override
 	public CompletionStage<Void> deleteFile(CloudPath file) {
-		return delegate.deleteFile(file) //
-				.whenComplete((nullReturn, exception) -> {
-					CompletionStage<CloudItemMetadata> future = CompletableFuture.failedFuture(new NotFoundException());
-					itemMetadataCache.put(file, future);
-					quotaCache.invalidateAll();
-				});
+		return delegate.deleteFile(file);
```

Comment:
no longer putting 404 response into the cache? previous implementation looked more efficient

Reply:
In the currently published version, as discussed, the cache is only used to remove duplicate requests (Request-Aggregation). In this case, the respective cloud implementation needs to handle better caching, which also includes 404.

In a subsequent commit, this cache is extended by a parameter that we can use to control whether only aggregation or real caching is used. In case of real caching the `expireAfterWrite` has to be set again and `whenComplete` looks a bit different (especially regarding this method, as you said, we will add again 404 responses but only if caching is required in this layer). Afterwards the cache handles both variants: Request-Aggregation and if necessary Caching.